kbfsgit/runner.go

```diff
@@ -579,62 +596,127 @@ func humanizeObjects(n int, d int) string {
+					elapsedStr := r.getElapsedStr(ctx, startTime, memProf, cpuProf)
+					// go-git seems to grab the lock during this phase. So we'd
+					// be flushing the journal before
+					// plumbing.StatusIndexOffset is done. So just skip
+					// printing "done." for this status here, but print it
+					// along with a newline when FSEventLock event is received
+					// in the case below.
+					if currStage != plumbing.StatusIndexOffset && !donePrinted {
+						r.errput.Write([]byte("done." + elapsedStr + "\n"))
+						donePrinted = true
+					}
 				}
+				if r.verbosity >= 4 {
+					cpuProf = filepath.Join(
+						os.TempDir(), fmt.Sprintf("cpu.%d.prof", update.Stage))
+					f, err := os.Create(cpuProf)
+					if err != nil {
+						r.log.CDebugf(
+							ctx, "Couldn't create CPU profile: %s", cpuProf)
+						cpuProf = ""
+					} else {
+						pprof.StartCPUProfile(f)
+					}
+				}
+				r.errput.Write([]byte(gogitStagesToStatus[update.Stage]))
+				donePrinted = false
+				lastByteCount = 0
+				currStage = update.Stage
+				startTime = r.config.Clock().Now()
+				r.log.CDebugf(ctx, "Entering stage: %s %s total",
+					gogitStagesToStatus[update.Stage], update.ObjectsTotal)
 			}
-			r.errput.Write([]byte(gogitStagesToStatus[update.Stage]))
-			lastByteCount = 0
-			currStage = update.Stage
-			startTime = r.config.Clock().Now()
-			r.log.CDebugf(ctx, "Entering stage: %s %s total",
-				gogitStagesToStatus[update.Stage], update.ObjectsTotal)
-		}
-		eraseStr := strings.Repeat("\b", lastByteCount)
-		newStr := ""
+			eraseStr := strings.Repeat("\b", lastByteCount)
+			newStr := ""
 
-		switch update.Stage {
-		case plumbing.StatusDone:
-			r.log.CDebugf(ctx, "Status processing done")
-			return
-		case plumbing.StatusCount:
-			newStr = fmt.Sprintf(
-				"%s objects... ", humanizeObjects(update.ObjectsTotal, 1))
-		case plumbing.StatusSort:
-		default:
-			newStr = fmt.Sprintf(
-				"(%.2f%%) %s objects... ",
-				percent(int64(update.ObjectsDone), int64(update.ObjectsTotal)),
-				humanizeObjects(update.ObjectsDone, update.ObjectsTotal))
-		}
+			switch update.Stage {
+			case plumbing.StatusDone:
+				r.log.CDebugf(ctx, "Status processing done")
+				return
+			case plumbing.StatusCount:
+				newStr = fmt.Sprintf(
+					"%s objects... ", humanizeObjects(update.ObjectsTotal, 1))
+			case plumbing.StatusSort:
+			default:
+				newStr = fmt.Sprintf(
+					"(%.2f%%) %s objects... ",
+					percent(int64(update.ObjectsDone), int64(update.ObjectsTotal)),
+					humanizeObjects(update.ObjectsDone, update.ObjectsTotal))
+			}
 
-		lastByteCount = len(newStr)
-		if r.progress {
-			r.errput.Write([]byte(eraseStr + newStr))
-		}
+			lastByteCount = len(newStr)
+			if r.progress {
+				r.errput.Write([]byte(eraseStr + newStr))
+			}
 
-		currStage = update.Stage
+			currStage = update.Stage
+		case fsEvent, ok := <-fsEvents:
+			if !ok {
+				fsEvents = nil
+				continue
+			}
+			switch fsEvent.EventType {
+			case libfs.FSEventLock:
```

Comment:
Isn't there data syncing during the Unlock event? It should be syncing the ref file and the parent dir blocks.  If you're worried about lots of `Close()` events triggering this, just move the `FSEvent` notification in `File` until after the lock is checked.

Reply:
That means we'd print two lines per pushed branch/ref now, right? Since `Lock` is where most of flushing (data blocks) happens, I figured it would be better to only do it for `Lock`. It would miss a few bytes being flushed, but we'd still miss some bytes at beginning even we print status for `Unlock` as well. WDYT?

Reply:
I just don't want any pauses in the output.  If the `Unlock()` flush happens to hit a disconnected or really slow network, I want the user to have some idea about why things are slow. If the second line really bothers you, maybe you can do something where the Unlock event is delayed by a second, and if `doneCh` is closed by the time the event arrives than nothing needs to be printed?  That should cover 99% of cases, since `Unlock` doesn't push much data.

Reply:
The delay approach sounds great. Will do!

Reply:
I ended up using the delay approach (with .5s instead of 1s) for both Lock and Unlock, so taht for multi-ref pushes on fast networks it only prints one line of bytes status. Let me know if you think it should only be for Lock!